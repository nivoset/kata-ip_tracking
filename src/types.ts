

export interface RequestHandled {
  (ipAddress: string): void;
}

export interface Top100 {
  (): string[];
}

export interface Clear {
  (): void;
}

export interface Tracker {
  requestHandled: RequestHandled;
  top100: Top100;
  clear: Clear;
}

export type TrackingData = [string, number];
export interface GetLowestEntry {
  (currentLow: TrackingData, entry: TrackingData): TrackingData;
}