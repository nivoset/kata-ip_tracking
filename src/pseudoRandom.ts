const sfc32 = (a: number, b: number, c: number, d: number) => () => {
    a >>>= 0; b >>>= 0; c >>>= 0; d >>>= 0; 
    var t = (a + b) | 0;
    a = b ^ b >>> 9;
    b = c + (c << 3) | 0;
    c = (c << 21 | c >>> 11);
    d = d + 1 | 0;
    t = t + d | 0;
    c = c + t | 0;
    return (t >>> 0) / 4294967296;
  };

export const pseudoRandomGenerator = (seed: number) => {
  return sfc32(0x9E3779B9, 0x243F6A88, 0xB7E15162, seed ^ 0xDEADBEEF);
}