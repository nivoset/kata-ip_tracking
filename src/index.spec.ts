import { generateTracker } from './index'
import { Tracker } from './types';
import { pseudoRandomGenerator } from './pseudoRandom'

let tracker: Tracker

beforeEach(() => {
  tracker = generateTracker();
});

it('should generate a new tracker with no records', () => {
  expect(tracker.top100()).toEqual([]);
})

it('should clear tracker when reset', () => {
  expect(tracker.top100()).toEqual([]);
  tracker.requestHandled("ip");

  expect(tracker.top100()).toEqual(["ip"]);
  tracker.clear();

  expect(tracker.top100()).toEqual([]);
})

it('should give top 100 results', () => {
  var rand = pseudoRandomGenerator(1337 ^ 0xDEADBEEF);
  expect(tracker.top100()).toMatchSnapshot();
  //generate 10000 pseudo-random requests and check if top 100 is correct
  const requests = [];
  for (let i = 0; i < 700; i++) {
    const r = Math.floor(rand() * 10000);
    requests.push([`ip index=${i} numberOfRequests=${r}`, r]);
  }
  requests.forEach(([ip, count]: any) => {
    new Array(count).fill(ip)
        .forEach(tracker.requestHandled)
  });

  expect(tracker.top100()).toMatchSnapshot();
})
