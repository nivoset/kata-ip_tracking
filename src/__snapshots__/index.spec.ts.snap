// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should give top 100 results 1`] = `Array []`;

exports[`should give top 100 results 2`] = `
Array [
  "ip 330 9982",
  "ip 532 9977",
  "ip 112 9965",
  "ip 339 9950",
  "ip 249 9917",
  "ip 674 9896",
  "ip 135 9877",
  "ip 108 9875",
  "ip 476 9841",
  "ip 600 9834",
  "ip 219 9811",
  "ip 338 9811",
  "ip 278 9792",
  "ip 300 9790",
  "ip 189 9784",
  "ip 173 9772",
  "ip 155 9741",
  "ip 673 9729",
  "ip 572 9722",
  "ip 687 9711",
  "ip 10 9710",
  "ip 676 9706",
  "ip 541 9705",
  "ip 619 9698",
  "ip 578 9692",
  "ip 418 9691",
  "ip 43 9672",
  "ip 152 9659",
  "ip 251 9651",
  "ip 158 9648",
  "ip 623 9647",
  "ip 92 9636",
  "ip 66 9631",
  "ip 157 9626",
  "ip 89 9615",
  "ip 148 9611",
  "ip 478 9607",
  "ip 586 9604",
  "ip 51 9577",
  "ip 651 9577",
  "ip 663 9561",
  "ip 337 9538",
  "ip 689 9519",
  "ip 693 9509",
  "ip 353 9508",
  "ip 396 9506",
  "ip 298 9500",
  "ip 24 9497",
  "ip 171 9484",
  "ip 435 9462",
  "ip 440 9432",
  "ip 232 9429",
  "ip 196 9411",
  "ip 456 9410",
  "ip 214 9400",
  "ip 294 9395",
  "ip 444 9366",
  "ip 428 9351",
  "ip 146 9308",
  "ip 288 9305",
  "ip 381 9281",
  "ip 143 9274",
  "ip 308 9255",
  "ip 265 9251",
  "ip 516 9229",
  "ip 56 9213",
  "ip 98 9213",
  "ip 81 9200",
  "ip 605 9180",
  "ip 318 9175",
  "ip 327 9165",
  "ip 144 9150",
  "ip 528 9142",
  "ip 178 9104",
  "ip 573 9102",
  "ip 420 9101",
  "ip 200 9099",
  "ip 597 9098",
  "ip 279 9062",
  "ip 59 9038",
  "ip 191 8990",
  "ip 625 8990",
  "ip 447 8962",
  "ip 622 8961",
  "ip 488 8952",
  "ip 127 8945",
  "ip 169 8945",
  "ip 183 8923",
  "ip 349 8910",
  "ip 128 8908",
  "ip 105 8897",
  "ip 477 8894",
  "ip 448 8886",
  "ip 139 8823",
  "ip 39 8817",
  "ip 79 8815",
  "ip 48 8813",
  "ip 667 8781",
  "ip 231 8774",
  "ip 553 8770",
]
`;
